### PR TITLE
Refactor conngen module to be a ConnBuilder

### PR DESCRIPTION
**Pull request by jougs**
This PR refactors the Connection Generator interface for NEST to be a regular `ConnBuilder` called `conngen` instead of being an extension module called `conngen`, which used to provide the `CGConnect` function. This addresses one of the action items of  #1717.

### CODE REVIEW

File: nestkernel/conn_builder_conngeninterface.cpp
Comment (jougs) [inline nestkernel/conn_builder_conngeninterface.cpp:118]: @hakonsbm: I was hoping that setting `weight_` and `delay_` here would have the effect that the subsequent call to `single_connect` will just use these values for establishing the connections. Unfortunately, this seems to not be the case and the connections are just created with defaults.

A completely orthogonal idea would of course be to directly call `ConnectionManager::connect()` here and not use any `Parameters` and `single_connect()` at all. For that I'd probably would need to get some values (e.g. the synapse model) from the *syn_spec*.

So, the point is that, the whole mechanism for `Parameters` and how to implement a `ConnBuilder` has apparently become a bit too complex for me and I'm a bit at loss here. Could you maybe give me some advice? I suggest to have a dedicated session on this during the hackathon next week in which we could also discuss a bit about how to document all this.

File: nestkernel/conn_builder_conngeninterface.cpp
Comment (jougs) [inline nestkernel/conn_builder_conngeninterface.cpp:76]: Remove to use defaults

File: doc/installation/compilation_options.rst
Comment (jougs) [inline doc/installation/compilation_options.rst:142]:
```suggestion
it provides support for the Connection Generator Interface from
```

File: nestkernel/conn_builder.cpp
Comment (jougs) [inline nestkernel/conn_builder.cpp:62]: @stinebuu: Not sure why I have this one. Maybe it should go away?

File: pynest/examples/csa_spatial_example.py
Comment (jougs) [inline pynest/examples/csa_spatial_example.py:44]:
```suggestion
       https://doi.org/10.3389/fninf.2014.00043
```

File: nestkernel/conn_builder.cpp
Comment (jougs) [reply nestkernel/conn_builder.cpp]: Removed in e40c193

File: nestkernel/conn_builder_conngen.cpp
Comment (stinebuu) [inline nestkernel/conn_builder_conngen.cpp:65]: You should probably also check if `syn_specs.size() > 1` and throw an error if this is the case. Collocation of synapses should not be possible if we use the `conngen` rule.

File: nestkernel/conn_builder_conngen.cpp
Comment (stinebuu) [inline nestkernel/conn_builder_conngen.cpp:62]: Is the indentation right here? It looks a bit weird, but it might just be me :)

File: nestkernel/conn_builder_conngen.cpp
Comment (stinebuu) [inline nestkernel/conn_builder_conngen.cpp:105]: It might be better to use `delay index and weight index` instead of `d_idx and w_idx` in the message.

**Review by stinebuu — CHANGES_REQUESTED** [review]
Nice! I have some comments and suggestions, but they shouldn't be too big.

File: examples/nest/conngen_interface.sli
Comment (jougs) [reply examples/nest/conngen_interface.sli:27]: Once upon a time, in a land far far away, we all believed that a C++ implementation of the CSA is on its way and will become available *very soon*. It was at that time that I created this file, which was meant to be a *test*. Then reality came (in total contrast to that C++ implementation) and I moved it to `examples` to keep it (hope dies last and all that...).

So much for history (and fairy tales). I have pushed a slightly updated version of the file in f844959be.

File: nestkernel/conn_builder_conngen.cpp
Comment (jougs) [reply nestkernel/conn_builder_conngen.cpp:105]: What about
```suggestion
      throw BadProperty( "The indices for weight and delay have to be either 0 or 1 and cannot be the same." );
```

File: nestkernel/conn_builder_conngen.cpp
Comment (jougs) [reply nestkernel/conn_builder_conngen.cpp]: Fixed in c2b4eade0.

File: nestkernel/conn_builder_conngen.cpp
Comment (jougs) [reply nestkernel/conn_builder_conngen.cpp:69]: Ah, yes. Now that you say it, I remember that we talked about this! Fixed in 824caee56. Could you maybe provide me with a code snipped that creates collocated synapses, so I can write a test that checks if the error reporting works? Thanks!

File: pynest/nest/tests/test_csa.py
Comment (jougs) [reply pynest/nest/tests/test_csa.py]: Much nicer, thanks! Fixed in ccfae0d05.

File: nestkernel/conn_builder_conngen.cpp
Comment (stinebuu) [reply nestkernel/conn_builder_conngen.cpp:109]: Looks good!

File: examples/nest/conngen_interface.sli
Comment (stinebuu) [reply examples/nest/conngen_interface.sli]: Makes sense, thanks for updating :)

File: nestkernel/conn_builder_conngen.cpp
Comment (stinebuu) [reply nestkernel/conn_builder_conngen.cpp:69]: A simple example would be:
```
node = nest.Create('iaf_psc_alpha', 3)
nest.Connect(node, node, syn_spec=nest.CollocatedSynapses({'weight': -2.}, {'weight': 2.}))
```

Comment (stinebuu) [reply nestkernel/conn_builder_conngen.cpp:69]: If you want to test without weight, you could use other parameters than weights, like `synapse_model` and other parameters that can be set for the individual synapse types.
```
node = nest.Create('iaf_psc_alpha', 3)
nest.Connect(node, node, syn_spec=nest.CollocatedSynapses({'synapse_model': 'stdp_synapse', 'alpha': 3.}, {'synapse_model': 'static_synapse'}))
```

**Review by sarakonradi — CHANGES_REQUESTED** [review]
Thanks! I've added some documentation-related suggestions inline.

File: doc/guides/connection_management.rst
Comment (jougs) [reply doc/guides/connection_management.rst]: Good idea. Done in d5802c9a5.

File: nestkernel/conn_builder_conngen.cpp
Comment (jougs) [reply nestkernel/conn_builder_conngen.cpp:69]: I've added a test in 500a7412b.

**Comment by jougs** [comment]
@stinebuu, @sarakonradi: I've addressed all your points in my recent commits. Not sure if we really need the review by @hakonsbm. If you think we're good without, feel free to say so or remove the request of review.

**Review by stinebuu — APPROVED** [review]
Thank you!

**Review by sarakonradi — APPROVED** [review]
LGTM!

**Pull request merged**
